### DISCUSSION TITLE
Improving Tasmota security document

### DISCUSSION THREAD

**Discussion by ebadi** [arendst/Tasmota#14166] [Ideas]
I read the Tasmota [security assumptions and scenarios](https://tasmota.github.io/docs/Securing-your-IoT-from-hacking/#disable-unsecured-fallback-wifi-wifimanager) and I would like to suggest few improvements.

1- As mentioned in the document, [decode_config](https://github.com/tasmota/decode-config) script can be used to decode the wifi AP/MQTT passwords that are stored in the device. Scenario 1 only consider the case in which the attacker has the WLAN key and therefore gain nothing by downloading the configuration file. However the following attackers are not considered:
- an attacker with access to router Ethernet ports
- an attacker that can exploit a node to ex-filtrate the configuration file. 
     - User browser's security protections (e.g. [Chrome's CORS-RFC1918](https://developer.chrome.com/blog/private-network-access-update/) ).  [see also](https://github.com/arendst/Tasmota/issues/6767)
     - [Server Side Request Forgery (SSRF)](https://owasp.org/www-community/attacks/Server_Side_Request_Forgery)

I therefore suggest to improve the documentation mentioning the importance of changing the [web admin password](https://github.com/arendst/Tasmota/pull/12900)

2- Improve the documentation to emphasis that [unsecured fallback](https://tasmota.github.io/docs/Securing-your-IoT-from-hacking/#disable-unsecured-fallback-wifi-wifimanager) is disabled by default (mine was set to 4 by default)

3- It is not clear what [CORS command](https://tasmota.github.io/docs/Commands/#wi-fi) does:
``` 
23:51:59.493 CMD: CORS test.com
23:51:59.501 MQT: stat/tasmota_0CC90E/RESULT = {"Command":"Unknown"}
23:52:10.657 CMD: CORS
23:52:10.664 MQT: stat/tasmota_0CC90E/RESULT = {"Command":"Unknown"}
```
I see these response headers:
![image](https://user-images.githubusercontent.com/1111940/147394433-ee757ecb-b48b-4c23-ac16-c16614c7957c.png)

**Comment by Jason2866** [comment]
There is **no security** for a esp8266 device if it can be accessed via Wlan or physical access.
Everything around done is mumbo jumbo.
To be clear the network where esp8266 devices are used needs to be secure!
Tasmota documentation is NOT about how to secure a network.

**Comment by barbudor** [comment]
@ebadi If you want to emphase the lack of security of ESP8266 device, please suggest docs update. You can do that with the pencil icon on the top right of the page.

Regarding point 3, CORS has NOTHING TO DO with security. It doesn't protect from anything making direct requests to URL.

**Comment by ebadi** [comment]
@barbudor @Jason2866 
> There is no security for a esp8266 device if it can be accessed via Wlan or physical access.

By physical ac, do you mean physical access to the chip or physical access to the network switch. There is a web admin page that protects a node with a **transitory** network access (via Ethernet port, browser) to get hold of wifi credentials that **permanently** endangers the **entire** wireless traffic (and not just my sonoff switch). So there are protections that are not well documented. 

> If you want to emphase the lack of security of ESP8266 device, please suggest docs update. 

No, I am first trying to **learn** before I suggest any update. As a new comer, I see the current security features (e.g. web admin password) can resolve the first issue. From your responses (@barbudor @Jason2866 ) it seems there are other issues that I am not aware of? It seems I am not qualified to make any change **yet**, so please share you thought and I will gladly update the document.

**Reply by barbudor** [reply]
If the device is connected by wifi, you first need to be connected to this wifi network before being able to use decode-config (or any other means) to sniff the wifi credentials.
This can be achieved only
- if you have already access to this Wifi network so you don't need the credentials anymore
- if you have a free access to a computer which is connected to the same network, so this is the failure point
If both case, the hacker is already in
I don't see any scenario where someone without access to the network can exploit Tasmota/ESP82666 to gain access to this network

CORS prevent a browser client-side Javascript to make Javascript web request to an URL outside of the server which serve the current page. So an attack would require that some external web site is compromised with code, so that when you  browse this web site, the malicious code tries to get access to your Tasmotas. It would not prevent any bot not running in the browser to seek for the same.

**Reply by ebadi** [reply]
> if you have already access to this Wifi network so you don't need the credentials anymore

Correct. 

> if you have a free access to a computer which is connected to the same network, so this is the failure point
If both case, the hacker is already in.

I disagree, please read the **transitory scenario** that I described earlier that gives the attacker access to wifi credentials and long term access to wifi network.

**Comment by Jason2866** [comment]
Any access to the device is unsecure. The webserver is built to be small and fast. There are for sure hundreds of possibilitys for exploits...  The underlying Nonos sdk is closed source, and has many known bugs and for sure a lot of unknown ones. You can not harden this device. You have to use in a secured very restricted (W)LAN.
A picture: See the device like a credit card. You have to take care of the access and usage.

**Reply by ebadi** [reply]
> There are for sure hundreds of possibilitys for exploits

Would it be possible for you to mention one?

> See the device like a credit card. You have to take care of the access and usage.

Where is your claim documented? You mean there is no point in protecting the web admin interface with password? So you mean all (currently over 2K) devices listed in [shodan](https://www.shodan.io/search?query=Tasmota) are vulnerable?

**Comment by Jason2866** [comment]
Yes, any esp8266 device is a potential risk. If someone with the needed skills invests enough time to hack.
That's a another reason why Tasmota exists. No Internet access no risk to be vulnerable.
In general every IoT device connected to Intenet is a big security problem. Is the code checked? How often do the get updates?

**Reply by ebadi** [reply]
> Yes, any esp8266 device is a potential risk. If someone with the needed skills invests enough time to hack.
> That's a another reason why Tasmota exists. No Internet access no risk to be vulnerable.
> In general every IoT device connected to Intenet is a big security problem. Is the code checked? How often do the get updates?

Thanks for sharing your opinion with me, but your claims are valid for any device on internet and not just IoT devices or Tasmota devices. So I don't really know how to respond. 

If what you are saying the official statement from the Tasmota project, the[Tasmota Securing your IoT from hacking](https://tasmota.github.io/docs/Securing-your-IoT-from-hacking/) and [Security Policy](https://github.com/arendst/Tasmota/blob/development/SECURITY.md) documents gives the users **the false sense of security**. @arendst  what is you take on this? do they need to be edited or just removed?

**Reply by Jason2866** [reply]
No conflict. Tasmota is designed to be used in a secured WLAN.
Read the article again. 
**Tasmota is NOT designed and build to be reachable from Internet.**
Security Policy is just a way to reach the maintainer for **general** security related stuff.
It was added because a used github actions had a security issue.

**Reply by ebadi** [reply]
> Read the article again.
>  Tasmota is NOT designed and build to be reachable from Internet.

Thanks, I read the document again, but I don't seem to find it. Am I reading the right document? which document and which part are you referring to?

**Reply by Jason2866** [reply]
https://github.com/arendst/Tasmota -> ABOUT
https://tasmota.github.io/docs/
`Total local control`

**Reply by Jason2866** [reply]
Nowhere is a connection to Internet mentioned nor SUGGESTED.

**Reply by sfromis** [reply]
**Any** device connected to the internet should be security-hardened. While no absolute security indeed exists, for devices/applications designed to be connected to the internet, "reasonable" precautions needs to be taken in development of hardware and software. This means that neither ESP8266 nor Tasmota is suitable for such usage. It works better on a WLAN, where you need suitable perimeter security, instead of having each and every device security-hardened on its own.

Thus, diving into security and exploit considerations around Tasmota may not be very useful. Yes, we know that it is not Fort Knox, and it does not try to be "really secure". While there are a few security features in Tasmota, like web admin password, these are not meant to be "real security", more like having a low-security lock to keep kids out of the candy cupboard.

**Reply by ebadi** [reply]
Thanks, 
I interpret `local control` as **it is not necessary** to be connected to internet/cloud (as an excellent feature) BUT not that **it should not**

**Reply by Jason2866** [reply]
It is up to you what you interpret. As said already nowhere it is mentioned Tasmota has the feature to be secure to be connected to the Internet.

**Reply by Jason2866** [reply]
It is normal for product descriptions if a feature is not mentioned that it is not there

**Reply by ebadi** [reply]
OK,  I see. an [IoT (**internet** of things) device](https://tasmota.github.io/docs/Securing-your-IoT-from-hacking/) that should not be connected to internet.

**Reply by sfromis** [reply]
As we say, the **S** in IoT means security.

**Comment by Jason2866** [comment]
One (known!) used security hole was TuyaConvert. This was used to update Tuya devices with a other firmware (for example Tasmota). Nobody knows how many devices are hacked from people with other things in mind...

**Comment by arendst** [comment]
Nice this shodan website! Searching for my IP address luckily only reveals my open HTTPS server and NOT my 60+ tasmota devices (which are behind the (default) firewall of course). Now I wish I only had patched my webserver with latest fixes....

**Comment by rapi3** [comment]
If you use a smart enough firewall ( opnsense / pfsese ) and you don't need to open ports for whole world you can restrict access by trusted IP, country IP or provider assignation.
That plus an IDS like Suricata/Snort and you can stop any IP that try to do a recognition scan before attack... so for my IPs shodan website will not find any port available to scan.

**Reply by s-hadinger** [reply]
Sorry but there is no such thing as a trusted IP or country IP. It is weak security and always possible to spoof or compromise. It's always a dangerous practice because people often overestimate the security features of IP addresses. In security you should consider that no IP is trusted. Period. 

TLS and encryption should be mandatory whenever you are exposed to the internet. This is where esp8266 is too limited to achieve high security. That's why we strongly suggest in Tasmota that those device are protected by perimetric security - which is unfortunately the only option left.

**Comment by ebadi** [comment]
Thanks everyone.
I opened [this](https://github.com/tasmota/docs/pull/882) pull request.

**Comment by ebadi** [comment]
I opened another pull request https://github.com/tasmota/docs/pull/884 on WifiConfig default value.
